### PR TITLE
Add a horrid Dockerfile

### PR DESCRIPTION
**Pull request by tianon**
```console
$ time docker build --no-cache .
Sending build context to Docker daemon 217.1 kB
Sending build context to Docker daemon 
Step 0 : FROM golang:1.4
 ---> ca0f230b927e
Step 1 : RUN mkdir -p /go/src/github.com/docker 	&& git clone --depth=1 https://github.com/docker/docker.git /go/src/github.com/docker/docker
 ---> Running in a06225dab76c
Cloning into '/go/src/github.com/docker/docker'...
 ---> 69a7c2d88901
Removing intermediate container a06225dab76c
Step 2 : ENV GOPATH $GOPATH:/go/src/github.com/docker/docker/vendor
 ---> Running in a08344598c3c
 ---> 8f6456696bf6
Removing intermediate container a08344598c3c
Step 3 : RUN cd /go/src/github.com/docker/docker 	&& bash hack/make/.go-autogen
 ---> Running in e0605f80d058
 ---> 59c6258fa160
Removing intermediate container e0605f80d058
Step 4 : RUN go get -d -v github.com/Sirupsen/logrus
 ---> Running in a12da22d30f9
 ---> 343385c95e0a
Removing intermediate container a12da22d30f9
Step 5 : WORKDIR /go/src/github.com/vbatts/docker-utils
 ---> Running in 03aad49b6584
 ---> dbe9f58a28e2
Removing intermediate container 03aad49b6584
Step 6 : COPY . /go/src/github.com/vbatts/docker-utils
 ---> 366e6ebb267a
Removing intermediate container 43a2b952c681
Step 7 : RUN go get -d -v ./...
 ---> Running in 0a9f6a9ad1d6
 ---> 284012f2f536
Removing intermediate container 0a9f6a9ad1d6
Step 8 : RUN go install -v ./...
 ---> Running in 8e56be6667f9
github.com/Sirupsen/logrus
github.com/docker/docker/pkg/tarsum
github.com/docker/docker/pkg/parsers
github.com/docker/docker/pkg/version
github.com/docker/libcontainer/user
github.com/docker/docker/pkg/ulimit
github.com/docker/docker/pkg/units
github.com/docker/docker/autogen/dockerversion
github.com/docker/docker/nat
github.com/docker/docker/pkg/term
github.com/docker/docker/pkg/timeutils
github.com/docker/docker/pkg/pubsub
github.com/docker/docker/pkg/ioutils
github.com/docker/docker/pkg/promise
github.com/docker/docker/pkg/system
github.com/docker/docker/graph/tags
github.com/docker/docker/pkg/reexec
github.com/docker/docker/daemon/network
github.com/docker/docker/pkg/homedir
github.com/docker/docker/pkg/parsers/filters
github.com/docker/docker/pkg/pools
github.com/docker/docker/pkg/stringid
github.com/tchap/go-patricia/patricia
github.com/docker/docker/pkg/fileutils
github.com/docker/docker/pkg/mflag
github.com/docker/docker/cliconfig
github.com/gorilla/context
github.com/docker/docker/pkg/parsers/kernel
github.com/docker/distribution/digest
github.com/docker/docker/pkg/archive
github.com/docker/docker/pkg/requestdecorator
github.com/docker/docker/pkg/timeoutconn
github.com/docker/libtrust
github.com/gorilla/mux
github.com/docker/docker/pkg/truncindex
github.com/docker/docker/pkg/urlutil
github.com/vbatts/docker-utils/sum
github.com/vbatts/docker-utils/version
github.com/vbatts/docker-utils/cmd/d2r/fsrv
github.com/docker/docker/opts
github.com/vbatts/docker-utils/dockerfile
github.com/vbatts/docker-utils/opts
github.com/vbatts/docker-utils/cmd/dockertarsum
github.com/docker/distribution/registry/api/v2
github.com/vbatts/docker-utils/cmd/docker-save-dockerfile
github.com/docker/docker/runconfig
github.com/docker/docker/daemon/graphdriver
github.com/docker/docker/pkg/chrootarchive
github.com/docker/docker/utils
github.com/docker/libtrust/trustgraph
github.com/docker/docker/pkg/jsonmessage
github.com/docker/docker/trust
github.com/docker/docker/daemon/events
github.com/docker/docker/pkg/httputils
github.com/docker/docker/pkg/streamformatter
github.com/docker/docker/api/types
github.com/docker/docker/image
github.com/docker/docker/pkg/progressreader
github.com/docker/docker/registry
github.com/docker/docker/graph
github.com/vbatts/docker-utils/registry
github.com/vbatts/docker-utils/cmd/docker-fetch
github.com/vbatts/docker-utils/cmd/d2r
 ---> 58be0d49775e
Removing intermediate container 8e56be6667f9
Successfully built 58be0d49775e

real	0m24.216s
user	0m0.022s
sys	0m0.029s
$ docker run -it --rm 58be0d49775e d2r -v
d2r - 1.0.1-dev
```

### CODE REVIEW

**Comment by tianon** [comment]
If Step 7 there ever has any output, it's time to adjust the `Dockerfile` with more pre-cache. :+1:

**Comment by tianon** [comment]
Full cache:
```console
$ time docker build .
Sending build context to Docker daemon 217.1 kB
Sending build context to Docker daemon 
Step 0 : FROM golang:1.4
 ---> ca0f230b927e
Step 1 : RUN mkdir -p /go/src/github.com/docker 	&& git clone --depth=1 https://github.com/docker/docker.git /go/src/github.com/docker/docker
 ---> Using cache
 ---> 69a7c2d88901
Step 2 : ENV GOPATH $GOPATH:/go/src/github.com/docker/docker/vendor
 ---> Using cache
 ---> 8f6456696bf6
Step 3 : RUN cd /go/src/github.com/docker/docker 	&& bash hack/make/.go-autogen
 ---> Using cache
 ---> 59c6258fa160
Step 4 : RUN go get -d -v github.com/Sirupsen/logrus
 ---> Using cache
 ---> 343385c95e0a
Step 5 : WORKDIR /go/src/github.com/vbatts/docker-utils
 ---> Using cache
 ---> dbe9f58a28e2
Step 6 : COPY . /go/src/github.com/vbatts/docker-utils
 ---> Using cache
 ---> 366e6ebb267a
Step 7 : RUN go get -d -v ./...
 ---> Using cache
 ---> 284012f2f536
Step 8 : RUN go install -v ./...
 ---> Using cache
 ---> 58be0d49775e
Successfully built 58be0d49775e

real	0m3.444s
user	0m0.022s
sys	0m0.022s
```

And now only changed some `.go` files:
```console
$ touch cmd/*/*.go
$ time docker build .
Sending build context to Docker daemon 217.1 kB
Sending build context to Docker daemon 
Step 0 : FROM golang:1.4
 ---> ca0f230b927e
Step 1 : RUN mkdir -p /go/src/github.com/docker 	&& git clone --depth=1 https://github.com/docker/docker.git /go/src/github.com/docker/docker
 ---> Using cache
 ---> 69a7c2d88901
Step 2 : ENV GOPATH $GOPATH:/go/src/github.com/docker/docker/vendor
 ---> Using cache
 ---> 8f6456696bf6
Step 3 : RUN cd /go/src/github.com/docker/docker 	&& bash hack/make/.go-autogen
 ---> Using cache
 ---> 59c6258fa160
Step 4 : RUN go get -d -v github.com/Sirupsen/logrus
 ---> Using cache
 ---> 343385c95e0a
Step 5 : WORKDIR /go/src/github.com/vbatts/docker-utils
 ---> Using cache
 ---> dbe9f58a28e2
Step 6 : COPY . /go/src/github.com/vbatts/docker-utils
 ---> 8d9cfe0139df
Removing intermediate container 772cef5a411b
Step 7 : RUN go get -d -v ./...
 ---> Running in 6a1322a793b2
 ---> 0c714f55ecd7
Removing intermediate container 6a1322a793b2
Step 8 : RUN go install -v ./...
 ---> Running in 0787b3d6ac77
github.com/Sirupsen/logrus
github.com/docker/docker/pkg/tarsum
github.com/docker/docker/pkg/parsers
github.com/docker/docker/pkg/version
github.com/docker/libcontainer/user
github.com/docker/docker/pkg/ulimit
github.com/docker/docker/pkg/units
github.com/docker/docker/autogen/dockerversion
github.com/docker/docker/pkg/term
github.com/docker/docker/pkg/timeutils
github.com/docker/docker/pkg/pubsub
github.com/docker/docker/pkg/ioutils
github.com/docker/docker/pkg/homedir
github.com/docker/docker/pkg/mflag
github.com/docker/docker/nat
github.com/docker/docker/cliconfig
github.com/docker/docker/pkg/promise
github.com/docker/docker/pkg/fileutils
github.com/docker/docker/pkg/pools
github.com/docker/docker/pkg/system
github.com/docker/distribution/digest
github.com/docker/docker/graph/tags
github.com/docker/docker/pkg/reexec
github.com/docker/docker/pkg/parsers/filters
github.com/docker/docker/daemon/network
github.com/docker/docker/pkg/stringid
github.com/tchap/go-patricia/patricia
github.com/gorilla/context
github.com/docker/docker/pkg/archive
github.com/docker/docker/pkg/parsers/kernel
github.com/docker/docker/pkg/requestdecorator
github.com/docker/docker/pkg/timeoutconn
github.com/docker/docker/pkg/truncindex
github.com/docker/docker/opts
github.com/docker/libtrust
github.com/gorilla/mux
github.com/docker/docker/pkg/urlutil
github.com/vbatts/docker-utils/sum
github.com/vbatts/docker-utils/version
github.com/vbatts/docker-utils/cmd/d2r/fsrv
github.com/vbatts/docker-utils/dockerfile
github.com/vbatts/docker-utils/opts
github.com/vbatts/docker-utils/cmd/dockertarsum
github.com/docker/docker/runconfig
github.com/vbatts/docker-utils/cmd/docker-save-dockerfile
github.com/docker/distribution/registry/api/v2
github.com/docker/docker/daemon/graphdriver
github.com/docker/docker/pkg/chrootarchive
github.com/docker/docker/pkg/jsonmessage
github.com/docker/docker/utils
github.com/docker/docker/api/types
github.com/docker/docker/daemon/events
github.com/docker/docker/image
github.com/docker/docker/pkg/httputils
github.com/docker/docker/pkg/streamformatter
github.com/docker/libtrust/trustgraph
github.com/docker/docker/pkg/progressreader
github.com/docker/docker/registry
github.com/docker/docker/trust
github.com/docker/docker/graph
github.com/vbatts/docker-utils/registry
github.com/vbatts/docker-utils/cmd/docker-fetch
github.com/vbatts/docker-utils/cmd/d2r
 ---> 01a80d01059a
Removing intermediate container 0787b3d6ac77
Successfully built 01a80d01059a

real	0m9.415s
user	0m0.024s
sys	0m0.028s
```

**Comment by vbatts** [comment]
Oh you

**Comment by tianon** [comment]
Also, :heart:

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.4
+
+# cache-fill before COPY
+# (docker/docker via git clone --depth=1 for speed)
+RUN mkdir -p /go/src/github.com/docker \
+	&& git clone --depth=1 https://github.com/docker/docker.git /go/src/github.com/docker/docker
+ENV GOPATH $GOPATH:/go/src/github.com/docker/docker/vendor
+# (also, naughty vbatts using things outside pkg/ from docker/docker)
+RUN cd /go/src/github.com/docker/docker \
+	&& bash hack/make/.go-autogen
```
Comment (vbatts) [inline Dockerfile]: great, isn't it? and folks still complain about GNU Makefiles...
/me sighs

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
+FROM golang:1.4
+
+# cache-fill before COPY
+# (docker/docker via git clone --depth=1 for speed)
+RUN mkdir -p /go/src/github.com/docker \
+	&& git clone --depth=1 https://github.com/docker/docker.git /go/src/github.com/docker/docker
+ENV GOPATH $GOPATH:/go/src/github.com/docker/docker/vendor
+# (also, naughty vbatts using things outside pkg/ from docker/docker)
```
Comment (vbatts) [inline Dockerfile]: i know. and it's bitten me too. perhaps we'll switch it to the docker/distribution soon

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+RUN cd /go/src/github.com/docker/docker \
+	&& bash hack/make/.go-autogen
+RUN go get -d -v github.com/Sirupsen/logrus
```
Comment (vbatts) [inline Dockerfile]: is this needed outside of the `go get ./...` below?

**Comment by vbatts** [comment]
p much LGTM

File: Dockerfile (hunk 3 above)
Comment (tianon) [inline Dockerfile]: That just makes the `go get` below super fast and no-op so you get shiny fast cache.  Also, if you add more deps later, the `go get` below makes sure they're covered.

Comment (vbatts) [inline Dockerfile]: sounds good

**Pull request merged**